The auto-update flag allows for regressions to sneak in

First of all, thanks for this tool! I'm working on migrating a moderately large code-base to full strict mode, and it will be instrumental in making that happen. I'm trying to figure out the best workflow for _when_, exactly, to run the CLI with the `auto-update` flag enabled. I want to make sure that files whose errors have all been fixed are removed from the loose files list as soon as possible, but here's a scenario that I just ran into:

1. Init the tool as per the instructions; receive a list of loosely-checked files and ignored errors
2. Work on one of those offending files to fix the errors
3. Run `loose-ts-check --auto-update` and see that the file in question is removed from the list 🎉 
4. Edit that file again to re-introduce a strict-based error
5. Run `loose-ts-check --auto-update` again

Here, the CLI correctly reports the error, but the `auto-update` mechanism then adds that file to the list of loosely-types files, which means running the compiler again will not produce the same error. The CLI reports this status:

```
error Command failed with exit code 2.                            
577 errors detected
576 errors have been ignored
1 errors were not ignored
1 errors could be ignored, as their error codes are ignored
src/data/config.ts(22,13): error TS2532: Object is possibly 'undefined'.
Additional 0 files will be ignored - registry will be updated
Updating the list of loosely type-checked files...
The list of loosely type-checked files updated successfully
```

Maybe I'm thinking about it incorrectly, but I would have expected that `auto-update` was a one-way street, only allowing files to be _removed_ rather than added. It seems like because that specific error _code_ was previously ignored, it allowed the offending file back onto the loose list.

For me, the question remains: how can I programmatically update (i.e., shrink) the list of loosely-checked files while ensuring that the files _not_ on that list don't secretly introduce regressions?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,25 @@
 
 ## Unreleased
 
+**BREAKING CHANGES**
+
+- `--auto-update` will no longer add new file paths to the registry
+
+  When there were errors that could be ignored because the error code is already
+  ignored in other files, `--auto-update` used to add these new file paths to
+  the registry. This allowed regressions to sneak in, since a file that was
+  already conforming to the stricter config could re-appear in the registry of
+  ignored files.
+
+  In situations like this one, these file paths will be no longer added to the
+  registry, and the program will fail when there are errors in files that could
+  be ignored.
+
+  If you need to start ignoring these errors, use the `--init` flag to
+  regenerate the registries.
+
+  See <https://github.com/Gelio/loose-ts-check/issues/16> for more information.
+
 ## v1.3.0 (2023-01-08)
 
 Features:

--- a/README.md
+++ b/README.md
@@ -17,7 +17,7 @@ incrementally, where only some existing files are allowed to have TS errors.
 - ignores specific TS errors from specific files
 - detects files that no longer have to be loosely type-checked
 - auto-updates the loosely type-checked list of files when a file no longer has
-  error
+  errors
 - auto-updates the ignored error codes when there are no errors of that type
 - loosely type-checked files can be specified using globs
 

--- a/src/cli/program.test.ts
+++ b/src/cli/program.test.ts
@@ -325,23 +325,6 @@ describe('program', () => {
         },
       };
 
-      it('should add new loosely type-checked files whose errors can be ignored', () => {
-        looselyTypeCheckedFiles = ['a', 'b'];
-        ignoredErrorCodes = ['TS1111', 'TS2222'];
-        const result = program(temporaryCliDependencies, [
-          createErrorLine('a', 'TS1111'),
-          createErrorLine('b', 'TS1111'),
-          createErrorLine('c', 'TS2222'),
-        ]);
-
-        expect(result).toBeUndefined();
-
-        expect(temporaryCliDependencies.saveJSONFile).toHaveBeenCalledWith(
-          temporaryCliDependencies.cliOptions['loosely-type-checked-files'],
-          ['a', 'b', 'c'],
-        );
-      });
-
       it('should remove loosely type-checked file paths that have no errors', () => {
         looselyTypeCheckedFiles = ['a', 'b'];
         ignoredErrorCodes = ['TS1111'];

--- a/src/cli/program.ts
+++ b/src/cli/program.ts
@@ -95,7 +95,6 @@ export const program = (
     reportTscErrorsThatCouldBeIgnored(
       cliDependencies,
       tscErrorsThatCouldBeIgnored,
-      updatedLooselyTypeCheckedFilePaths,
     ),
     reportLooselyTypeCheckedFilePathsWithoutErrors(
       cliDependencies,

--- a/src/cli/reporting/report-tsc-errors-that-could-be-ignored.test.ts
+++ b/src/cli/reporting/report-tsc-errors-that-could-be-ignored.test.ts
@@ -1,10 +1,8 @@
-import { getDefaultCliOptions } from '../get-default-cli-options';
 import { reportTscErrorsThatCouldBeIgnored } from './report-tsc-errors-that-could-be-ignored';
 
 describe('reportTscErrorsThatCouldBeIgnored', () => {
   const dependencies: Parameters<typeof reportTscErrorsThatCouldBeIgnored>[0] =
     {
-      cliOptions: getDefaultCliOptions(),
       log: jest.fn(),
     };
 
@@ -13,54 +11,20 @@ describe('reportTscErrorsThatCouldBeIgnored', () => {
   });
 
   it('should not fail when there are no errors', () => {
-    const result = reportTscErrorsThatCouldBeIgnored(
-      dependencies,
-      [],
-      new Set(),
-    );
+    const result = reportTscErrorsThatCouldBeIgnored(dependencies, []);
 
     expect(result).toBeUndefined();
   });
 
-  it('should fail when there are errors and auto-update is disabled', () => {
-    const result = reportTscErrorsThatCouldBeIgnored(
-      dependencies,
-      [
-        {
-          filePath: 'a',
-          rawErrorLines: [],
-          tscErrorCode: 'TS1234',
-        },
-      ],
-      new Set(),
-    );
+  it('should fail when there are errors', () => {
+    const result = reportTscErrorsThatCouldBeIgnored(dependencies, [
+      {
+        filePath: 'a',
+        rawErrorLines: [],
+        tscErrorCode: 'TS1234',
+      },
+    ]);
 
     expect(result?.shouldFail).toBe(true);
-  });
-
-  it('should update the loosely type-checked files when auto-update is on', () => {
-    const temporaryDependencies: typeof dependencies = {
-      ...dependencies,
-      cliOptions: {
-        ...dependencies.cliOptions,
-        'auto-update': true,
-      },
-    };
-
-    const looselyTypeCheckedFiles = new Set<string>();
-    const result = reportTscErrorsThatCouldBeIgnored(
-      temporaryDependencies,
-      [
-        {
-          filePath: 'a',
-          rawErrorLines: [],
-          tscErrorCode: 'TS1234',
-        },
-      ],
-      looselyTypeCheckedFiles,
-    );
-
-    expect(result).toBeUndefined();
-    expect(looselyTypeCheckedFiles.has('a')).toBe(true);
   });
 });

--- a/src/cli/reporting/report-tsc-errors-that-could-be-ignored.ts
+++ b/src/cli/reporting/report-tsc-errors-that-could-be-ignored.ts
@@ -4,9 +4,8 @@ import { CliDependencies } from '../cli-dependencies';
 import { ReportResult } from './report-result';
 
 export function reportTscErrorsThatCouldBeIgnored(
-  { log, cliOptions }: Pick<CliDependencies, 'log' | 'cliOptions'>,
+  { log }: Pick<CliDependencies, 'log'>,
   tscErrorsThatCouldBeIgnored: TscError[],
-  looselyTypeCheckedFilePaths: Set<string>,
 ): ReportResult {
   if (tscErrorsThatCouldBeIgnored.length === 0) {
     return;
@@ -22,29 +21,13 @@ export function reportTscErrorsThatCouldBeIgnored(
     log(tscError.rawErrorLines.join('\n')),
   );
 
-  if (!cliOptions['auto-update']) {
-    log(
-      `Use the ${green(
-        '--auto-update',
-      )} option to update the registry automatically.`,
-    );
-
-    return {
-      shouldFail: true,
-    };
-  }
-
-  const initialLooselyTypeCheckedFilesCount =
-    tscErrorsThatCouldBeIgnored.length;
-  tscErrorsThatCouldBeIgnored.forEach((tscError) => {
-    looselyTypeCheckedFilePaths.add(tscError.filePath);
-  });
-  const additionalLooselyTypeCheckedFilePaths =
-    tscErrorsThatCouldBeIgnored.length - initialLooselyTypeCheckedFilesCount;
-
   log(
-    `Additional ${yellow(
-      additionalLooselyTypeCheckedFilePaths,
-    )} files will be ignored - registry will be updated`,
+    `Use the ${green(
+      '--init',
+    )} option to add these file paths to the registry.`,
   );
+
+  return {
+    shouldFail: true,
+  };
 }

